service/tablet_allocator.cc

```diff
@@ -2532,11 +2532,48 @@ class tablet_allocator_impl : public tablet_allocator::impl
                       table, tablets.tablet_count(), new_tablets.tablet_count());
         co_return std::move(new_tablets);
     }
+
+    // The merging of tablet is completely based on the power-of-two constraint.
+    // Tablet of ids X and X+1 are merged into new tablet id (X >> 1).
+    future<tablet_map> merge_tablets(token_metadata_ptr tm, table_id table) {
+        auto& tablets = tm->tablets().get_tablet_map(table);
+
+        tablet_map new_tablets(tablets.tablet_count() / 2);
+
+        for (tablet_id tid : new_tablets.tablet_ids()) {
+            co_await coroutine::maybe_yield();
+
+            tablet_id old_left_tid = tablet_id(tid.value() << 1);
+            tablet_id old_right_tid = tablet_id(old_left_tid.value() + 1);
+
+            auto& left_tablet_info = tablets.get_tablet_info(old_left_tid);
+            auto& right_tablet_info = tablets.get_tablet_info(old_right_tid);
+
+            auto sorted = [] (tablet_replica_set set) {
+                std::ranges::sort(set, std::less<tablet_replica>());
+                return set;
+            };
+            auto left_tablet_replicas = sorted(left_tablet_info.replicas);
+            auto right_tablet_replicas = sorted(right_tablet_info.replicas);
+            if (left_tablet_replicas != right_tablet_replicas) {
+                throw std::runtime_error(format("Sibling tablets {} (r: {}) and {} (r: {}) are not colocated.",
+                                                old_left_tid, left_tablet_replicas, old_right_tid, right_tablet_replicas));
+            }
+
+            new_tablets.set_tablet(tid, left_tablet_info);
```

Comment:
Ah I forget to address this by not throwing away current tablet info, but rather preserving its fields when building new tablet info.

Reply:
We cannot merge tasks in general, since if left and right might have different task ids and then there's no way to merge them - we can store only one task per tablet. So merge finalization should proceed only if all tasks are mergeable. We can start with something simple and not merge if there are any tasks.

Reply:
Ah I remember when you raised this to me, we were going with a timestamp or something, so I'd take a minimum among infos, but apparently we changed that.

Reply:
There is still `repair_time` which we should merge, yes. `repair_task_info` is another problem.

Reply:
why we cannot merge repair tasks if tablets are co-located? it should be fine as long as repair works with token ranges, and doesn't require tablet id stability. I am proceeding with this constraint, but let me know it indeed exists.

Reply:
I don't know exactly how repair tasks are implemented, but we shouldn't have any active repair task when state machine transition into resize finalization state, no?

Reply:
anyway, I will only emit the request to finalize when all sibling tablets have their infos mergeable.

Reply:
> why we cannot merge repair tasks if tablets are co-located? it should be fine as long as repair works with token ranges, and doesn't require tablet id stability. I am proceeding with this constraint, but let me know it indeed exists.

left tablet may have a different task id assigned than the right tablet, so they belong to different jobs. Merging would mean that one of the tasks would be dropped and no-oped, which is not correct. The dropped task may be considered as finished without doing any work.

Reply:
> I don't know exactly how repair tasks are implemented, but we shouldn't have any active repair task when state machine transition into resize finalization state, no?

We shouldn't have active repair transitions, but we may still have inactive (to be scheduled) tasks assigned. They can be assigned at any point in time in group0 history, even after state machine entered split finalization.

Reply:
> anyway, I will only emit the request to finalize when all sibling tablets have their infos mergeable.

OK